beacon_node/network/src/beacon_processor/worker/gossip_methods.rs

```diff
@@ -204,76 +430,185 @@ impl<T: BeaconChainTypes> Worker<T> {
+
+        self.process_gossip_aggregate_result(
+            result,
+            beacon_block_root,
+            message_id,
+            peer_id,
+            reprocess_tx,
+            seen_timestamp,
+        );
+    }
+
+    pub fn process_gossip_aggregate_batch(
+        self,
+        packages: Vec<GossipAggregatePackage<T::EthSpec>>,
+        reprocess_tx: Option<mpsc::Sender<ReprocessQueueMessage<T>>>,
+    ) {
+        let aggregates = packages.iter().map(|package| package.aggregate.as_ref());
+
+        let results = match self
+            .chain
+            .batch_verify_aggregated_attestations_for_gossip(aggregates)
+        {
+            Ok(results) => results,
+            Err(e) => {
+                error!(
+                    self.log,
+                    "Batch agg. attn verification failed";
+                    "error" => ?e
                 );
                 return;
             }
         };
 
-        // Indicate to the `Network` service that this message is valid and can be
-        // propagated on the gossip network.
-        self.propagate_validation_result(message_id, peer_id, MessageAcceptance::Accept);
+        // Sanity check.
+        if results.len() != packages.len() {
+            // The log is `crit` since in this scenario we might be penalizing/rewarding the wrong
+            // peer.
+            crit!(
+                self.log,
+                "Batch agg. attestation result mismatch";
+                "results" => results.len(),
+                "packages" => packages.len(),
+            )
+        }
 
-        // Register the attestation with any monitored validators.
-        self.chain
-            .validator_monitor
-            .read()
-            .register_gossip_aggregated_attestation(
-                seen_timestamp,
-                aggregate.aggregate(),
-                aggregate.indexed_attestation(),
-                &self.chain.slot_clock,
+        // Map the results into a new `Vec` so that `results` no longer holds a reference to
+        // `packages`.
+        #[allow(clippy::needless_collect)] // The clippy suggestion fails the borrow checker.
+        let results = results
+            .into_iter()
+            .map(|result| result.map(|verified| verified.into_indexed_attestation()))
+            .collect::<Vec<_>>();
+
+        for (result, package) in results.into_iter().zip(packages.into_iter()) {
+            let result = match result {
+                Ok(indexed_attestation) => Ok(VerifiedAggregate {
+                    indexed_attestation,
+                    signed_aggregate: *package.aggregate,
+                }),
+                Err(error) => Err(RejectedAggregate {
+                    signed_aggregate: *package.aggregate,
+                    error,
+                }),
+            };
+
+            self.process_gossip_aggregate_result(
+                result,
+                package.beacon_block_root,
+                package.message_id.clone(),
```

Comment:
```suggestion
                package.message_id,
```

(_not_ part of https://github.com/michaelsproul/lighthouse/commit/fe3921d50d7f7367652e9419ce7842deb1a05f84)